File: baekjoon/others/01347-cartography.py
from enum import Enum


class Direction(Enum):
    UP = 0
    RIGHT = 1
    DOWN = 2
    LEFT = 3


if __name__ == '__main__':
    maze = [[False for _ in range(101)] for _ in range(101)]

    n = int(input())

    move_str = input().strip()

    row = 50
    col = 50
    min_row = row
    max_row = row
    min_col = col
    max_col = col
    maze[row][col] = True

    direction = Direction.DOWN

    for c in move_str:
        if c == 'F':
            if direction == Direction.UP:
                row -= 1
                min_row = min(min_row, row)
            elif direction == Direction.RIGHT:
                col += 1
                max_col = max(max_col, col)
            elif direction == Direction.DOWN:
                row += 1
                max_row = max(max_row, row)
            elif direction == Direction.LEFT:
                col -= 1
                min_col = min(min_col, col)
            maze[row][col] = True
        elif c == 'L':
            direction = Direction((direction.value + 3) % 4)
        elif c == 'R':
            direction = Direction((direction.value + 1) % 4)

    for i in range(min_row, max_row + 1):
        for j in range(min_col, max_col + 1):
            print('.' if maze[i][j] is True else '#', end='')
        print()

File: baekjoon/simulation-only/01347-cartography.cpp
#include <iostream>
#include <vector>

using namespace std;

enum class Direction : int { UP = 0, RIGHT, DOWN, LEFT };

size_t width = 1;
int row = 0, col = 0;
vector<vector<bool>> maze{{true}};

vector<bool> make_line_vec() {
    vector<bool> line_vec;
    line_vec.resize(width, false);
    return line_vec;
}

void move_on() {
    if (row == -1) {
        // Negative row
        maze.insert(maze.begin(), make_line_vec());
        row = 0;
    } else if (maze.size() < static_cast<size_t>(row + 1)) {
        // Not enough row
        maze.resize(row + 1, make_line_vec());
    }
    if (col == -1) {
        // Negative col
        width++;
        for (auto& row_vec : maze) {
            row_vec.insert(row_vec.begin(), false);
        }
        col = 0;
    } else if (maze[row].size() < static_cast<size_t>(col + 1)) {
        // Not enough col
        if (width < static_cast<size_t>(col + 1)) {
            width = col + 1;
        }
        maze[row].resize(width, false);
    }
    maze[row][col] = true;
}

void print_maze() {
    for (const auto& row_vec : maze) {
        for (const auto cell : row_vec) {
            cout << (cell ? '.' : '#');
        }
        for (size_t i = row_vec.size(); i < width; i++) {
            cout << '#';
        }
        cout << '\n';
    }
}

int main() {
    // Get n
    int n;
    cin >> n;

    // Move on
    auto direction = Direction::DOWN;
    while (n > 0) {
        int c = cin.get();
        switch (c) {
        case 'F':
            switch (direction) {
            case Direction::UP:
                row--;
                break;
            case Direction::RIGHT:
                col++;
                break;
            case Direction::DOWN:
                row++;
                break;
            case Direction::LEFT:
                col--;
            }
            move_on();
            break;
        case 'L':
            direction = static_cast<Direction>((static_cast<int>(direction) + 3) % 4);
            break;
        case 'R':
            direction = static_cast<Direction>((static_cast<int>(direction) + 1) % 4);
            break;
        default:
            continue;
        }
        n--;
    }

    print_maze();

    return 0;
}
